src/azure-cli/fish-shell/create_fish_complete.py

```diff
@@ -0,0 +1,134 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import subprocess
+import re
+
+
+def fix_strings(text: str):
+    return text.replace("*", "").replace("'", "").replace('"', "")
+
+
+# Write to local az.fish
+with open("az.fish", "w") as f:
+    subcommands = [['az']]
+
+    # Match on lines that we care about (lines that are tabbed and :)
+    command_lines_re = re.compile(r'^\s{4}(?!az)\w+(\w|-)+.*')
+    # Match linescontaing a flag
+    flag_lines_re = re.compile(r'^\s{4}--(\w|-)+.*')
+    # Parse out commands (eg az --help will parse out account, acr, etc)
+    command_re = re.compile(r'\w+(\w|-)+(?=\s+:)')
+    # Parse out command flags (eg --help)
+    flag_re = re.compile(r'--(\w|-)+(?=\s+:)')
+    # Parse out descriptions (anything after :)
+    description_re = re.compile(r'(?<=:\s).*$')
+
+    output = []
+    did_write = False
+
+    # Create a function to match on a specific command
+    f.write('''
+function __get_cmd
+    for x in $argv
+        if string match -v -q -- '-*' $x
+            echo $x
+        else
+            break
+        end
+    end
+end
+
+function __is_az_subcommand
+    set -l cmd (commandline -poc)
+    set -l cmds (__get_cmd $cmd[2..])
+    set -l args (string join -- ' ' $argv)
+    set -l sub (string join -- ' ' $cmds)
+    if [ "$sub" = "$args" ]
+        return 0
+    end
+    return 1
+end
+
+''')
+
+    is_az = True
+
+    while len(subcommands) != 0:
+
+        # Grab last item in queue (DFS)
+        command = subcommands.pop()
+        # Query back help page
+        response = subprocess.run(command + ['--help'], stdout=subprocess.PIPE).stdout.decode('utf-8')
```

Comment:
How is the performance? How about we get the candidates in advance and build the index just like modern IDE does.

Reply:
Performance of the python script is not great. It takes on average 20m to generate the completions. The reason I went this route is it requires no code changes when a module or its help doc is changed. Looking at the `help.py` for each command module might be significantly faster but looks like wed still need to do some parsing.

That said once the completions are generated they are faster so I would say its a compile vs runtime speed. And this script only needs to be run once where itll generate a `az.fish` file the user can source.

Here is a snippet of what the python script generates: 
```sh
complete -c az -d "Manage Azure subscription information." -n "__fish_az_no_subcommand" -f -a "account"
complete -c az -d "Manage private registries with Azure Container" -n "__fish_az_no_subcommand" -f -a "acr"
complete -c az -d "Manage Azure Active Directory Graph entities needed" -n "__fish_az_no_subcommand" -f -a "ad"
complete -c az -d "Manage Azure Advisor." -n "__fish_az_no_subcommand" -f -a "advisor"
complete -c az -d "Manage Azure Kubernetes Services." -n "__fish_az_no_subcommand" -f -a "aks"
complete -c az -d "Manage Azure Media Services resources." -n "__fish_az_no_subcommand" -f -a "ams"
complete -c az -d "Manage Azure API Management services." -n "__fish_az_no_subcommand" -f -a "apim"
```

That said if we are concerned about "compile-time" speed, do we have the help docs in a better format that the `help.py` (eg. yaml, json, etc)

Reply:
@mattpasco-microsoft Oh! Sorry, I thought `-h` is running at runtime. I like your solution. We run this file only once.

Reply:
Correct this will only run once, ideally in the build time!